File: src/okf_parser/git_commit.py
"""Subject-first OKF commit-message parsing and formatting."""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from types import MappingProxyType
from typing import Mapping

from okf_parser.digests import canonical_json, normalize_newlines, parsed_digest, source_digest
from okf_parser.models import YamlValue
from okf_parser.parser import DocumentParseError, parse_document_text

_RESERVED_KEYS = frozenset(
    {
        "title",
        "source_kind",
        "repository_identity",
        "object_format",
        "object_type",
        "oid",
        "commit_oid",
        "tree_oid",
        "parent_oids",
        "parents",
        "author",
        "committer",
        "refs",
        "source_digest",
        "parsed_digest",
        "diagnostics",
        "provenance",
    }
)


class GitCommitMessageError(ValueError):
    """Report a structural error in a subject-first OKF commit message."""

    def __init__(self, code: str, message: str, *, line: int | None = None) -> None:
        super().__init__(message)
        self.code = code
        self.line = line


@dataclass(frozen=True, slots=True)
class GitCommitMessage:
    """Canonical projection of one valid UTF-8 Git commit message."""

    subject: str
    authored_metadata: Mapping[str, YamlValue]
    effective_frontmatter: Mapping[str, YamlValue]
    body: str
    has_envelope: bool
    source_digest: str
    parsed_digest: str

    @property
    def concept_type(self) -> str:
        """Return the effective semantic OKF type."""
        value = self.effective_frontmatter["type"]
        assert isinstance(value, str)
        return value


def _parse_metadata(block: str) -> dict[str, YamlValue]:
    """Reuse the strict filesystem YAML authority without inventing Git YAML semantics."""
    source = f"---\n{block}\n---\n"
    try:
        parsed = parse_document_text(Path("<git-commit-metadata>"), source)
    except DocumentParseError as exc:
        raise GitCommitMessageError("GIT_MESSAGE_YAML", str(exc), line=3) from exc
    return dict(parsed.frontmatter)


def _validate_metadata(metadata: dict[str, YamlValue]) -> str:
    reserved = sorted(set(metadata) & _RESERVED_KEYS)
    if reserved:
        joined = ", ".join(reserved)
        raise GitCommitMessageError(
            "GIT_MESSAGE_RESERVED_KEY",
            f"commit metadata uses adapter-owned key(s): {joined}",
            line=3,
        )

    if "type" not in metadata:
        return "Commit"
    raw_type = metadata["type"]
    if not isinstance(raw_type, str) or not raw_type.strip():
        raise GitCommitMessageError(
            "GIT_MESSAGE_TYPE",
            "commit metadata type must be a non-empty string",
            line=3,
        )
    return raw_type.strip()


def _find_closing_delimiter(envelope_source: str) -> tuple[str, str, int]:
    """Return metadata block, post-close text and the closing line number."""
    lines = envelope_source.splitlines(keepends=True)
    if not lines or lines[0].removesuffix("\n") != "--- okf":
        raise GitCommitMessageError(
            "GIT_MESSAGE_ENVELOPE",
            "structured commit metadata must start with exact '--- okf'",
            line=3,
        )

    offset = len(lines[0])
    for index, line in enumerate(lines[1:], start=1):
        logical = line.removesuffix("\n")
        if logical == "---":
            metadata = envelope_source[len(lines[0]) : offset]
            after = envelope_source[offset + len(line) :]
            return metadata.removesuffix("\n"), after, index + 3
        offset += len(line)

    raise GitCommitMessageError(
        "GIT_MESSAGE_UNTERMINATED_ENVELOPE",
        "structured commit metadata has no closing '---' delimiter",
        line=3,
    )


def parse_git_commit_message(source: str) -> GitCommitMessage:
    """Parse a valid UTF-8 commit message into its effective OKF projection."""
    exact_digest = source_digest(source)
    normalized = normalize_newlines(source).removeprefix("\ufeff")
    subject, separator, remainder = normalized.partition("\n")
    if not subject.strip():
        raise GitCommitMessageError(
            "GIT_MESSAGE_EMPTY_SUBJECT",
            "commit subject must be non-empty",
            line=1,
        )

    has_blank_separator = separator == "\n" and remainder.startswith("\n")
    candidate = remainder[1:] if has_blank_separator else remainder
    has_envelope = candidate == "--- okf" or candidate.startswith("--- okf\n")

    metadata: dict[str, YamlValue]
    body: str
    if has_envelope:
        block, after, closing_line = _find_closing_delimiter(candidate)
        if after and not after.startswith("\n"):
            raise GitCommitMessageError(
                "GIT_MESSAGE_BODY_SEPARATOR",
                "structured commit metadata must be followed by a blank line before the body",
                line=closing_line + 1,
            )
        body = after[1:] if after.startswith("\n") else ""
        if any(line == "--- okf" for line in body.splitlines()):
            raise GitCommitMessageError(
                "GIT_MESSAGE_DUPLICATE_ENVELOPE",
                "commit message contains more than one OKF envelope",
            )
        metadata = _parse_metadata(block)
    else:
        metadata = {}
        body = candidate if has_blank_separator else remainder

    effective_type = _validate_metadata(metadata)
    effective: dict[str, YamlValue] = dict(metadata)
    effective["type"] = effective_type
    effective["title"] = subject

    return GitCommitMessage(
        subject=subject,
        authored_metadata=MappingProxyType(dict(metadata)),
        effective_frontmatter=MappingProxyType(effective),
        body=body,
        has_envelope=has_envelope,
        source_digest=exact_digest,
        parsed_digest=parsed_digest(effective, body),
    )


def validate_git_commit_message(
    source: str,
    *,
    require_envelope: bool = False,
) -> GitCommitMessage:
    """Validate one prospective commit message and return its canonical projection."""
    parsed = parse_git_commit_message(source)
    if require_envelope and not parsed.has_envelope:
        raise GitCommitMessageError(
            "GIT_MESSAGE_ENVELOPE_REQUIRED",
            "repository policy requires an OKF commit envelope",
            line=3,
        )
    return parsed


def _format_metadata(metadata: Mapping[str, YamlValue]) -> str:
    """Emit a deterministic JSON-flow subset that is also valid failsafe YAML."""
    return "\n".join(
        f"{canonical_json(key)}: {canonical_json(metadata[key])}"
        for key in sorted(metadata, key=lambda value: value.encode("utf-16-be"))
    )


def format_git_commit_message(message: GitCommitMessage) -> str:
    """Format one parsed commit message idempotently without rewriting Git history."""
    if not message.has_envelope:
        return message.subject if not message.body else f"{message.subject}\n\n{message.body}"

    metadata = _format_metadata(message.authored_metadata)
    block = "--- okf\n"
    if metadata:
        block += f"{metadata}\n"
    block += "---"
    return f"{message.subject}\n\n{block}" if not message.body else f"{message.subject}\n\n{block}\n\n{message.body}"

File: typescript/src/git-commit.ts
import { parseDocumentContent } from "./core.js";
import type { FrontmatterValue } from "./core.js";
import { canonicalJson, normalizeNewlines, parsedDigest, sourceDigest } from "./digests.js";

const RESERVED_KEYS = new Set([
  "title",
  "source_kind",
  "repository_identity",
  "object_format",
  "object_type",
  "oid",
  "commit_oid",
  "tree_oid",
  "parent_oids",
  "parents",
  "author",
  "committer",
  "refs",
  "source_digest",
  "parsed_digest",
  "diagnostics",
  "provenance",
]);

export class GitCommitMessageError extends Error {
  readonly code: string;
  readonly line?: number;

  constructor(code: string, message: string, line?: number) {
    super(message);
    this.name = "GitCommitMessageError";
    this.code = code;
    if (line !== undefined) this.line = line;
  }
}

export interface GitCommitMessage {
  readonly subject: string;
  readonly authoredMetadata: Readonly<Record<string, FrontmatterValue>>;
  readonly effectiveFrontmatter: Readonly<Record<string, FrontmatterValue>>;
  readonly body: string;
  readonly hasEnvelope: boolean;
  readonly sourceDigest: string;
  readonly parsedDigest: string;
  readonly conceptType: string;
}

function parseMetadata(block: string): Readonly<Record<string, FrontmatterValue>> {
  try {
    return parseDocumentContent(`---\n${block}\n---\n`, "<git-commit-metadata>").frontmatter;
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    throw new GitCommitMessageError("GIT_MESSAGE_YAML", message, 3);
  }
}

function validateMetadata(metadata: Readonly<Record<string, FrontmatterValue>>): string {
  const reserved = Object.keys(metadata)
    .filter((key) => RESERVED_KEYS.has(key))
    .sort();
  if (reserved.length > 0) {
    throw new GitCommitMessageError(
      "GIT_MESSAGE_RESERVED_KEY",
      `commit metadata uses adapter-owned key(s): ${reserved.join(", ")}`,
      3,
    );
  }

  if (!("type" in metadata)) return "Commit";
  const rawType = metadata.type;
  if (typeof rawType !== "string" || rawType.trim() === "") {
    throw new GitCommitMessageError(
      "GIT_MESSAGE_TYPE",
      "commit metadata type must be a non-empty string",
      3,
    );
  }
  return rawType.trim();
}

function findClosingDelimiter(envelopeSource: string): readonly [string, string, number] {
  const lines = envelopeSource.match(/.*(?:\n|$)/gu)?.filter((line) => line !== "") ?? [];
  const first = lines[0]?.endsWith("\n") ? lines[0].slice(0, -1) : lines[0];
  if (first !== "--- okf") {
    throw new GitCommitMessageError(
      "GIT_MESSAGE_ENVELOPE",
      "structured commit metadata must start with exact '--- okf'",
      3,
    );
  }

  let offset = lines[0]?.length ?? 0;
  for (let index = 1; index < lines.length; index += 1) {
    const line = lines[index] ?? "";
    const logical = line.endsWith("\n") ? line.slice(0, -1) : line;
    if (logical === "---") {
      const metadata = envelopeSource.slice(lines[0]?.length ?? 0, offset).replace(/\n$/u, "");
      const after = envelopeSource.slice(offset + line.length);
      return [metadata, after, index + 3] as const;
    }
    offset += line.length;
  }

  throw new GitCommitMessageError(
    "GIT_MESSAGE_UNTERMINATED_ENVELOPE",
    "structured commit metadata has no closing '---' delimiter",
    3,
  );
}

export function parseGitCommitMessage(source: string): GitCommitMessage {
  const exactDigest = sourceDigest(source);
  const normalized = normalizeNewlines(source).replace(/^\uFEFF/u, "");
  const newline = normalized.indexOf("\n");
  const subject = newline < 0 ? normalized : normalized.slice(0, newline);
  const remainder = newline < 0 ? "" : normalized.slice(newline + 1);
  if (subject.trim() === "") {
    throw new GitCommitMessageError("GIT_MESSAGE_EMPTY_SUBJECT", "commit subject must be non-empty", 1);
  }

  const hasBlankSeparator = newline >= 0 && remainder.startsWith("\n");
  const candidate = hasBlankSeparator ? remainder.slice(1) : remainder;
  const hasEnvelope = candidate === "--- okf" || candidate.startsWith("--- okf\n");

  let metadata: Readonly<Record<string, FrontmatterValue>>;
  let body: string;
  if (hasEnvelope) {
    const [block, after, closingLine] = findClosingDelimiter(candidate);
    if (after !== "" && !after.startsWith("\n")) {
      throw new GitCommitMessageError(
        "GIT_MESSAGE_BODY_SEPARATOR",
        "structured commit metadata must be followed by a blank line before the body",
        closingLine + 1,
      );
    }
    body = after.startsWith("\n") ? after.slice(1) : "";
    if (body.split("\n").some((line) => line === "--- okf")) {
      throw new GitCommitMessageError(
        "GIT_MESSAGE_DUPLICATE_ENVELOPE",
        "commit message contains more than one OKF envelope",
      );
    }
    metadata = parseMetadata(block);
  } else {
    metadata = Object.freeze({});
    body = hasBlankSeparator ? candidate : remainder;
  }

  const effectiveType = validateMetadata(metadata);
  const effectiveFrontmatter = Object.freeze({ ...metadata, type: effectiveType, title: subject });
  return Object.freeze({
    subject,
    authoredMetadata: Object.freeze({ ...metadata }),
    effectiveFrontmatter,
    body,
    hasEnvelope,
    sourceDigest: exactDigest,
    parsedDigest: parsedDigest(effectiveFrontmatter, body),
    conceptType: effectiveType,
  });
}

export function validateGitCommitMessage(
  source: string,
  options: { readonly requireEnvelope?: boolean } = {},
): GitCommitMessage {
  const parsed = parseGitCommitMessage(source);
  if (options.requireEnvelope === true && !parsed.hasEnvelope) {
    throw new GitCommitMessageError(
      "GIT_MESSAGE_ENVELOPE_REQUIRED",
      "repository policy requires an OKF commit envelope",
      3,
    );
  }
  return parsed;
}

function formatMetadata(metadata: Readonly<Record<string, FrontmatterValue>>): string {
  return Object.keys(metadata)
    .sort()
    .map((key) => `${canonicalJson(key)}: ${canonicalJson(metadata[key] ?? null)}`)
    .join("\n");
}

export function formatGitCommitMessage(message: GitCommitMessage): string {
  if (!message.hasEnvelope) {
    return message.body === "" ? message.subject : `${message.subject}\n\n${message.body}`;
  }

  const metadata = formatMetadata(message.authoredMetadata);
  let block = "--- okf\n";
  if (metadata !== "") block += `${metadata}\n`;
  block += "---";
  return message.body === ""
    ? `${message.subject}\n\n${block}`
    : `${message.subject}\n\n${block}\n\n${message.body}`;
}
